src/common/QueryHandler.ts

```diff
@@ -191,11 +210,60 @@ const pierceHandler = createInternalQueryHandler({
   },
 });
 
-const builtInHandlers = new Map([
-  ['aria', ariaHandler],
-  ['pierce', pierceHandler],
+const xpathHandler = internalizeCustomQueryHandler({
+  queryOne: (element, selector) => {
+    const doc = element.ownerDocument || document;
+    const result = doc.evaluate(
+      selector,
+      element,
+      null,
+      XPathResult.FIRST_ORDERED_NODE_TYPE
+    );
+    return result.singleNodeValue;
+  },
+
+  queryAll: (element, selector) => {
+    const doc = element.ownerDocument || document;
+    const iterator = doc.evaluate(
+      selector,
+      element,
+      null,
+      XPathResult.ORDERED_NODE_ITERATOR_TYPE
+    );
+    const array: Node[] = [];
+    let item;
+    while ((item = iterator.iterateNext())) {
+      array.push(item);
+    }
+    return array;
+  },
+});
+
+const transformXPathSelector = (selector: string): string => {
+  if (selector.startsWith('//')) {
```

Comment:
I think we only need this transformation if the waitForSelector is invoked on the element handle

Reply:
This is just a noop at Page level.

Reply:
why does the test reports .// in the error message then? https://github.com/puppeteer/puppeteer/pull/8730/files#diff-20e9a9715adedba9766c324afd9350bd8187dd7e166cbbd43888a561623d6f83L738

I know that the selectors are equivalent but I don't think we should be rewriting them without a need.

Reply:
Because the element doesn't exist. I don't think it makes sense to handle page handle logic separately, given this was also previous behavior.

Reply:
Hm, I don't think it was the previous behavior. Previously, the selector would be only rewritten when called on an ElementHandle.